From 0ac7e5148a9dbcf4713ec93a2e5d5506d4b5622a Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 28 Oct 2025 15:45:07 +0800
Subject: [PATCH 090/105] mtk: mt76: mt7996: allow setting antenna with
 radio_idx

Add the support to set antenna of a specified radio.
Note that the input antenna should not be shifted.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 mt7996/main.c | 33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/mt7996/main.c b/mt7996/main.c
index 5a6acb868..40d87e452 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2040,6 +2040,36 @@ mt7996_set_coverage_class(struct ieee80211_hw *hw, int radio_idx,
 	mutex_unlock(&dev->mt76.mutex);
 }
 
+static int
+mt7996_set_radio_antenna(struct ieee80211_hw *hw, int radio_idx, u32 tx_ant, u32 rx_ant)
+{
+	struct mt7996_dev *dev = mt7996_hw_dev(hw);
+	struct mt7996_phy *phy;
+	struct wiphy_radio *radio;
+	int shift;
+
+	if (radio_idx >= hw->wiphy->n_radio)
+		return -EINVAL;
+
+	mutex_lock(&dev->mt76.mutex);
+
+	phy = dev->radio_phy[radio_idx];
+	radio = &dev->radios[radio_idx];
+	shift = dev->mt76.chainshift[phy->mt76->band_idx];
+
+	/* tx_ant is not shifted */
+	phy->mt76->chainmask = tx_ant & (radio->antenna_mask >> shift);
+	phy->mt76->antenna_mask = phy->mt76->chainmask & phy->orig_antenna_mask;
+
+	mt76_set_stream_caps(phy->mt76, true);
+	mt7996_set_stream_vht_txbf_caps(phy);
+	mt7996_set_stream_he_eht_caps(phy);
+
+	mutex_unlock(&dev->mt76.mutex);
+
+	return 0;
+}
+
 static int
 mt7996_set_antenna(struct ieee80211_hw *hw, int radio_idx,
 		   u32 tx_ant, u32 rx_ant)
@@ -2052,6 +2082,9 @@ mt7996_set_antenna(struct ieee80211_hw *hw, int radio_idx,
 	if (tx_ant != rx_ant)
 		return -EINVAL;
 
+	if (radio_idx != -1)
+		return mt7996_set_radio_antenna(hw, radio_idx, tx_ant, rx_ant);
+
 	for (i = 0; i < hw->wiphy->n_radio; i++) {
 		phy = dev->radio_phy[i];
 		radio = &dev->radios[i];
-- 
2.45.2

